Frontend_Web_Development_React_Angular_Vue/React/Week 2/2.1_Comparing Functional and Class Components.md

```diff
@@ -0,0 +1,288 @@
+
+## Unmounting
+
+The next phase in the lifecycle is when a component is removed from the DOM, or _unmounting_ as React likes to call it.
+
+React has only one built-in method that gets called when a component is unmounted:
+
+-   `componentWillUnmount()`
+
+> **componentWillUnmount**
+
+The `componentWillUnmount` method is called when the component is about to be removed from the DOM.
+
+## What is State (in React)?
+
+This is a very crucial topic to know about, before you know what stateful and stateless components actually are.
+
+State stores data about the component that can change over time. Change could come in the form of user events or system events such as response to user input or server requests. Working with a state’s component normally involves setting a component’s default state, accessing the current state and updating the state.
+
+> An example would be shown soon, shortly
+
+![React Hooks Logo](https://spycoding.com/wp-content/uploads/2020/10/react_hooks.png)
+
+## What are React Hooks?
+
+Hooks is an additional API that is coming in the limelight with functional components. It is a blazing fire amongst the React ecosystem since the `v16.8` public release.
+
+React Hooks are in-built functions that allow React developers to use state and lifecycle methods inside functional components, they also work together with existing code, so they can easily be adopted into a codebase. The way Hooks were pitched to the public was that they allow developers to use state in functional components but under the hood, Hooks are much more powerful than that. They allow React Developers to enjoy the following benefits:
+
+-   Improved code reuse;
+-   Better code composition;
+-   Better defaults;
+-   Sharing non-visual logic with the use of custom hooks;
+-   Flexibility in moving up and down the `components` tree.
+
+With React Hooks, developers get the power to use functional components for almost everything they need to do from just rendering UI to also handling state and also logic — which is pretty neat.
+
+**To put things simple. Consider Hooks API provide a state to functional components (as they are stateless, being non-object oriented, not wrapped in a class).**
+
+Well, by now you must choose your way, between **class** and **functional** components approach.
+
+Coming to the example as told above, I would like to implement a **button clicked** state storage using both the approaches. This shall enable you to choose as per your comfort.
+
+```jsx
+import React from "react";
+
+class App extends React.Component {
+	constructor(props) {
+		super(props); // Inherited child classes should call parent's constructor.
+		this.state = {
+			buttonClicked: false,
+			clickedTimes: 0,
+			avgClickDuration: 0,
+		};
+	}
+	// render() prints the values in the state
+	render() {
+		let val = ``;
+		for (const key in this.state) val += `${key} : ${this.state[key]}, `;
+		return <div>{val}</div>;
+	}
+}
+
+export default App;
+```
```

Comment:
This definitely works but it seems like a hack to make this work. While learning react someone shouldnt come across this  kind of code. I have made a small change to this. Have a look at it: https://codesandbox.io/s/nifty-sea-e590x?file=/src/App.js

Reply:
> This definitely works but it seems like a hack to make this work. While learning react someone shouldnt come across this kind of code. I have made a small change to this. Have a look at it: https://codesandbox.io/s/nifty-sea-e590x?file=/src/App.js

### REVIEW ACCEPTED

The [code snippet ](https://codesandbox.io/s/nifty-sea-e590x?file=/src/App.js) was carefully seen and seemed much better and easy to me (though being a bit lengthy) and thus I have decided to accept your suggestion and has been reflected finally.

Thank you,
for showing such deep concern into the issue, and showing great support while contributing.